SDL: Fix mouse capture in HiDPI mode

kalkie opened the pull request:
This PR fixes two independent, but very similar issues in how the mouse is captured to the drawable area in HiDPI mode. Both are needed for proper mouse control in scaled HiDPI mode, when the drawable area doesn't fit the whole window. I was experiencing the issues on a MacBook with Retina display, when the ScummVM graphics mode was set to OpenGL.

### SDL: Fix mouse clip to game area in HiDPI mode

The mouse position is set in window coordinates, but it's clipped to the game area in drawable area coordinates. Previously, the scaling between these two was not taken into account when calculating the right/bottom edges of the game area. When the clipped mouse position was converted back to window coordinates and rounded to the nearest integer, it could end up on the edge of the game area, not inside of it. This leads to a loop in which the clipped mouse position is outside of the game area and needs to be clipped again.

https://bugs.scummvm.org/ticket/12646
https://bugs.scummvm.org/ticket/13075 (likely duplicate)

### SDL: Fix SDL_SetWindowMouseRect in HiDPI mode

In HiDPI mode, the window coordinates and the drawable area coordinates might be on a different scale. The allowed mouse area needs to be scaled accordingly before passing it to `SDL_SetWindowMouseRect()`.

This issue is likely the same as:
https://bugs.scummvm.org/ticket/13152
The description of that ticket is not very detailed, but the symptoms sounds very similar.

kalkie commented on backends/platform/sdl/sdl-window.cpp:
```diff
@@ -168,10 +168,11 @@ void SdlWindow::grabMouse(bool grab) {
 }
 
 void SdlWindow::setMouseRect(const Common::Rect &rect) {
```
I'm not exactly sure in which coordinates (window/drawable area) the parameter is expected to be in. Thus, I followed what was done in `SdlWindow::warpMouseInWindow()`, which expects drawable area coordinates, although its documentation refers to window coordinates. Maybe the terminology is not fully harmonized after the HiDPI mode was introduced.

kalkie commented:
@criezy, it looks like you have been working on HiDPI scaling and mouse related issues in the past. If you have the time, would you be willing to take a look at this?

criezy commented:
For the first issue, I can see that there is indeed an issue with the code, but I am wondering if commit 618350e is the best fix for it. It seems to me that the issue is not the clipping done in `SdlGraphicsManager::notifyMousePosition()` but rather the rounding done in `SdlWindow::warpMouseInWindow()`. It indeed seems to me that it should be rounding down and not to the nearest there.
As an example, if the scaling is 2, the first two pixels of a row (x = 0 and x = 1) correspond to the first non-scaled pixel. So x = 1 should be rounded to 0 when we undo the scaling. But this is currently rounded to 1.

And fixing that rounding would also fix the issue I think?

criezy commented:
I think the logic for the second commit (d2f6b34) is probably correct. This code was added fairly recently in pull request #3546, so maybe @ccawley2011 has some insight. But I suspect `SDL_SetWindowMouseRect` is another case where SDL is not consistent between platforms and expect non-scaled coordinates on macOS and scaled coordinates on Windows (and possibly Linux) on a HiDPI display, and using `getSdlDpiScalingFactor()` takes care of this discrepancy.

However, as I noted above for your first commit. I think we might need to round down and not round to nearest after dividing by the scaling factor. Although here that probably doesn't matter in practice.

kalkie commented:
Thanks for the comments, @criezy.

> For the first issue, I can see that there is indeed an issue with the code, but I am wondering if commit [618350e](https://github.com/scummvm/scummvm/commit/618350eb2e7bdfb90c526c995228368c346f6da1) is the best fix for it. It seems to me that the issue is not the clipping done in `SdlGraphicsManager::notifyMousePosition()` but rather the rounding done in `SdlWindow::warpMouseInWindow()`. It indeed seems to me that it should be rounding down and not to the nearest there. As an example, if the scaling is 2, the first two pixels of a row (x = 0 and x = 1) correspond to the first non-scaled pixel. So x = 1 should be rounded to 0 when we undo the scaling. But this is currently rounded to 1.
> 
> And fixing that rounding would also fix the issue I think?

I didn't want to change the way the coordinate conversions are currently done. The current way of rounding to nearest in both directions makes the conversion symmetric, like it should be. If we change the conversion from drawable to window coordinates round down, then the window to drawable conversion should round up to make the operations symmetric. Unfortunately, at least alone, these changes do not fix the issue. This was actually the first thing I tried, but it only moved the problem to the left hand side of the screen. Depending on the scaling settings, the edges of the game's viewport might not be in coordinates that are divisible by the scaling factor. For example, with a scaling factor of 2.0, a 4:3 game with fit to window scaling has the game's left side edge on drawable coordinates 213 on my screen. If this is converted to window coordinates by rounding down, the result will not be within the drawable area anymore.

One solution I was thinking about was to do the mouse clipping always in window coordinates, as that's the coordinate space we get the mouse position in the first place. In other words, if the drawable area is first converted to window coordinates, and the check and possible clipping would be done there, this kind of conversion errors would not be possible. I didn't want to refactor more than I needed to fix the bug, and thus I didn't proceed this further. I can give it a try if you think it makes more sense.

But before we go there, I'll give a few words in the defence of the current approach in this PR 😄 Before the HiDPI support, the mouse position and the `_activeArea.drawRect` were both in the same coordinate space. Because `struct Rect` has its `bottom` and `right` edges exclusive, i.e., not part of the rectangle, the clipping had to take this into account my subtracting one pixel from those. This same thing applies still, but this one pixel in window coordinates might nowadays be a different amount of pixels in the drawable area coordinates. My change takes this into account by calculating how many pixels in the drawable area correspond to one pixel in the window coordinates.

criezy commented:
I indeed forgot that the game viewport might not start from pixel 0 and that changing the rounding would become an issue when it starts from a pixel that is not a multiple of the dpi scaling. If somebody has the courage it could be interesting to work out how to properly convert the positions back and forth to make sure we stamp out all the edge cases out current code may have. It might involve shifting before dividing by the scaling, rounding and then shifting back for example. I am not sure the code we have would work properly for cases where the dpi scaling might be 3 for example, or, even worse, a non-integral value. But as far as I know those are only hypothetical for now.

For now the solution you propose makes sense and should be sufficient. I find however that it is not clear why we subtract `(int)(dpiScaling + 0.5f)`. If somebody revisit this code in the future he might struggle to understand the logic. It could be a bit clearer with `(int)(1 * dpiScaling + 0.5f)`, but I think adding a comment could be a better idea.

kalkie commented:
> For now the solution you propose makes sense and should be sufficient. I find however that it is not clear why we subtract `(int)(dpiScaling + 0.5f)`. If somebody revisit this code in the future he might struggle to understand the logic. It could be a bit clearer with `(int)(1 * dpiScaling + 0.5f)`, but I think adding a comment could be a better idea.

I agree, and did both, actually. I changed the conversions to `(int)(1 * dpiScale + 0.5f)` so that they match visually with other similar conversions. This way they could be easier to spot if one would add functions for these coordinate conversions at some point. Also added the comment trying to explain why this is needed, as it's likely not clear without knowing the current issue.

criezy reviewed (approved):
That looks good to merge now. Thank you for your work on this issue.
